File: src/shared/stateless/dotnet/Boutique.Stateless/Retry.cs
// Copyright 2026 Google LLC
// Licensed under the Apache License, Version 2.0 (the "License");

using StackExchange.Redis;

namespace Boutique.Stateless;

public enum RetryClass
{
    Permanent,
    Conflict,
    Dependency
}

public static class Retry
{
    public static RetryClass Classify(Exception exception)
    {
        ArgumentNullException.ThrowIfNull(exception);
        if (exception is AggregateConflictException)
        {
            return RetryClass.Conflict;
        }
        if (exception is OperationCanceledException or ArgumentException)
        {
            return RetryClass.Permanent;
        }
        if (exception is RedisConnectionException or RedisTimeoutException)
        {
            return RetryClass.Dependency;
        }
        var message = exception.Message.ToUpperInvariant();
        if (new[] { "MOVED ", "ASK ", "TRYAGAIN", "CLUSTERDOWN", "LOADING ",
                    "CONNECTION RESET", "CONNECTION REFUSED", "I/O TIMEOUT" }
            .Any(message.Contains))
        {
            return RetryClass.Dependency;
        }
        if (new[] { "WRONG LAST SEQUENCE", "WRONG LAST SEQ", "REVISION MISMATCH" }
            .Any(message.Contains))
        {
            return RetryClass.Conflict;
        }
        return RetryClass.Permanent;
    }

    public static TimeSpan Backoff(int attempt, TimeSpan minimum, TimeSpan maximum)
    {
        attempt = Math.Max(attempt, 0);
        if (minimum <= TimeSpan.Zero)
        {
            minimum = TimeSpan.FromMilliseconds(1);
        }
        if (maximum < minimum)
        {
            maximum = minimum;
        }
        var capMilliseconds = minimum.TotalMilliseconds;
        for (var step = 0; step < attempt && capMilliseconds < maximum.TotalMilliseconds; step++)
        {
            capMilliseconds = Math.Min(capMilliseconds * 2, maximum.TotalMilliseconds);
        }
        return TimeSpan.FromMilliseconds(Random.Shared.NextDouble() * capMilliseconds);
    }
}

File: src/shared/stateless/go/retry.go
// Copyright 2026 Google LLC
// Licensed under the Apache License, Version 2.0 (the "License");

package stateless

import (
	"context"
	"errors"
	"math/rand/v2"
	"net"
	"strings"
	"time"
)

type RetryClass uint8

const (
	RetryPermanent RetryClass = iota
	RetryConflict
	RetryDependency
)

// ClassifyRetry gives handlers one shared distinction between conflicts,
// transient dependency failures, and poison/permanent input errors.
func ClassifyRetry(err error) RetryClass {
	if err == nil {
		return RetryPermanent
	}
	if errors.Is(err, ErrConflict) || errors.Is(err, ErrLeaseHeld) ||
		errors.Is(err, ErrLeaseLost) {
		return RetryConflict
	}
	if errors.Is(err, context.DeadlineExceeded) {
		return RetryDependency
	}
	if errors.Is(err, context.Canceled) || errors.Is(err, ErrInvalidCommit) ||
		errors.Is(err, ErrLeaseComplete) {
		return RetryPermanent
	}
	var temporary interface{ Temporary() bool }
	if errors.As(err, &temporary) && temporary.Temporary() {
		return RetryDependency
	}
	var networkError net.Error
	if errors.As(err, &networkError) {
		return RetryDependency
	}
	message := strings.ToUpper(err.Error())
	for _, marker := range []string{
		"MOVED ", "ASK ", "TRYAGAIN", "CLUSTERDOWN", "LOADING ",
		"CONNECTION RESET", "CONNECTION REFUSED", "I/O TIMEOUT",
	} {
		if strings.Contains(message, marker) {
			return RetryDependency
		}
	}
	for _, marker := range []string{
		"WRONG LAST SEQUENCE", "WRONG LAST SEQ", "REVISION MISMATCH",
	} {
		if strings.Contains(message, marker) {
			return RetryConflict
		}
	}
	return RetryPermanent
}

// Backoff returns capped exponential delay with full jitter.
func Backoff(attempt int, minimum, maximum time.Duration) time.Duration {
	if attempt < 0 {
		attempt = 0
	}
	if minimum <= 0 {
		minimum = time.Millisecond
	}
	if maximum < minimum {
		maximum = minimum
	}
	capDelay := minimum
	for step := 0; step < attempt && capDelay < maximum; step++ {
		if capDelay > maximum/2 {
			capDelay = maximum
		} else {
			capDelay *= 2
		}
	}
	if capDelay > maximum {
		capDelay = maximum
	}
	if capDelay <= 1 {
		return capDelay
	}
	return time.Duration(rand.Int64N(int64(capDelay) + 1))
}
